src/peripheral/scb.rs

```diff
@@ -518,21 +555,80 @@ impl SCB {
         crate::asm::isb();
     }
 
+    /// Invalidates an object from the D-cache
+    ///
+    /// * `obj`: Object to invalidate
+    ///
+    /// Invalidates cache starting from the lowest 32-byte aligned address containing `obj`,
+    /// in blocks of 32 bytes until all of `obj` has been invalidated.
+    ///
+    /// Invalidation causes the next read access to memory to be fetched from main memory instead
+    /// of the cache.
+    ///
+    /// # Safety
+    ///
+    /// After invalidating, `obj` will be read from main memory on next access. This may cause
+    /// recent writes to `obj` to be lost, potentially including the write that initialised it.
+    /// Therefore, this method may cause uninitialised memory or invalid values to be read,
+    /// resulting in undefined behaviour. You must ensure that main memory contains a valid and
+    /// initialised value for T before invalidating `obj`.
```

Comment:
Isn't it unsound in general when the memory a `&T` points to is modified (unless `T` is an `UnsafeCell`)?

Reply:
Do you think we should require `&mut` here instead of `&`, or just give a more dire warning about potential consequences?

Reply:
`&mut` sounds like a good option, as long as it doesn't make the API unusable. Do you happen to have any real-world examples where this API would be used?

Reply:
I've used the previous invalidate API in ethernet peripheral drivers where the descriptors and packet memory are shared between CPU and peripheral. In all of those instances the code had &mut access, though the receive buffers were in principle read-only.

The other use case I can see is sharing memory between cores on a multi-core device, e.g. the STM32H7 dual-core series have D-cache for the M7 core, so communicating with the M4 core could require cache management. Imagine a synchronisation primitive which needs to invalidate some interior memory to ensure it's reading a fresh value from the other core, and wants to be able to do this with only `&self`. But probably such a primitive should contain an `UnsafeCell` which can handle the interior mutability, and then it would still be able to pass &mut to the cache invalidation function.

On balance I think it's probably a more obvious API if it takes `&mut` to make it clear that the value might change, and in the rare case someone needs to access it with a non-mut, it's only a one line function that calls another public unsafe method anyway, so it's not that much extra effort. I'll update this PR.